### PR TITLE
Adds new_nuget_package

### LINKED ISSUE #26: bazel run csharp_binary: mono not found
Running a csharp_binary fails to find mono:

```sh
pcj:~/github/rules_protobuf/examples/helloworld/csharp$ bazel run GreeterClient
INFO: Found 1 target...
Target //examples/helloworld/csharp/GreeterClient:GreeterClient up-to-date:
  bazel-bin/examples/helloworld/csharp/GreeterClient/GreeterClient.exe
  bazel-bin/examples/helloworld/csharp/GreeterClient/GreeterClient
INFO: Elapsed time: 0.084s, Critical Path: 0.00s

INFO: Running command line: bazel-bin/examples/helloworld/csharp/GreeterClient/GreeterClient
/private/var/tmp/_bazel_pcj/63330772b4917b139280caef8bb81867/execroot/rules_protobuf/bazel-out/local-fastbuild/bin/examples/helloworld/csharp/GreeterClient/GreeterClient: line 14: org_pubref_rules_protobuf/external/mono/bin/mono: No such file or directory
```

```sh
#!/bin/bash

if [[ -e "$0.runfiles" ]]; then
  cd $0.runfiles/org_pubref_rules_protobuf
fi

# TODO(jeremy): This is a gross and fragile hack.
# We should be able to do better than this.
ln -s -f org_pubref_rules_protobuf/examples/helloworld/csharp/GreeterClient/GreeterClient.exe $(basename examples/helloworld/csharp/GreeterClient/GreeterClient.exe)
for l in ../nuget_google_protobuf/Google.Protobuf.3.0.0/lib/net45/Google.Protobuf.dll ../nuget_google_protobuf/Google.Protobuf.3.0.0/lib/netstandard1.0/Google.Protobuf.dll ../nuget_grpc/Grpc.Core.1.0.0/lib/net45/Grpc.Core.dll examples/helloworld/csharp/Greeter/Greeter.dll ../nuget_google_protobuf/Google.Protobuf.3.0.0/lib/net45/Google.Protobuf.dll ../nuget_google_protobuf/Google.Protobuf.3.0.0/lib/netstandard1.0/Google.Protobuf.dll ../nuget_grpc/Grpc.Core.1.0.0/lib/net45/Grpc.Core.dll; do
    ln -s -f org_pubref_rules_protobuf/$l $(basename org_pubref_rules_protobuf/$l)
done

org_pubref_rules_protobuf/external/mono/bin/mono $(basename examples/helloworld/csharp/GreeterClient/GreeterClient.exe) "$@"
```

### PR DESCRIPTION
**Pull request by pcj**
- Adds new_nuget_package as workaround for Issue #18.
- csharp_binary fixes for path outputs. Closes #26.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by pcj** [comment]
@zaphar or @damienmg Can you review?  These two pull requests make it possible to build csharp grpc applications with https://github.com/pubref/rules_protobuf.  I'd like to make rules_csharp a prerequisite for `csharp_proto_library` in that repo.

```python
new_nuget_package(
    name = "nuget_grpc",
    package = "Grpc",
    version = "1.0.0",
    build_file_content =
"""
load("@io_bazel_rules_dotnet//dotnet:csharp.bzl", "dll_import")
dll_import(
  name = "runtime_osx",
  srcs = glob(["Grpc.Core.1.0.0/runtimes/osx/**/*.dll"]),
  visibility = ["//visibility:public"],
)
dll_import(
  name = "core",
  srcs = glob(["Grpc.Core.1.0.0/lib/net45/**/*.dll"]),
  visibility = ["//visibility:public"],
)
"""
)
```

```sh
pcj:~/github/rules_protobuf$ bazel run examples/helloworld/csharp/GreeterClient
INFO: Found 1 target...
Target //examples/helloworld/csharp/GreeterClient:GreeterClient up-to-date:
  bazel-bin/examples/helloworld/csharp/GreeterClient/GreeterClient.exe
  bazel-bin/examples/helloworld/csharp/GreeterClient/GreeterClient
INFO: Elapsed time: 16.289s, Critical Path: 0.00s

INFO: Running command line: bazel-bin/examples/helloworld/csharp/GreeterClient/GreeterClient
Greeting: Hello you
Press any key to exit...
```

**Comment by damienmg** [comment]
Jenkins, test this please

File: dotnet/csharp.bzl [hunk 1]
```diff
@@ -359,7 +361,7 @@ Args:
 
 csharp_binary = rule(
     implementation = _csc_compile_impl,
-    attrs = dict(_COMMON_ATTRS.items() + _EXE_ATTRS.items()),
```
Comment (damienmg) [inline dotnet/csharp.bzl]: do not revert those change, they are because dictionary addition is not python compatible so this file can be parsed as python

Comment (pcj) [inline dotnet/csharp.bzl]: OK, I didn't realize that was a requirement.  Will fix that back.

Comment (pcj) [inline dotnet/csharp.bzl]: Done with 3158c2d054bae77297e22361ee6660e5714702b3

File: dotnet/csharp.bzl [hunk 2]
```diff
@@ -529,4 +579,17 @@ def csharp_repositories(use_local_mono=False):
       # work when Workspaces import this using a repository rule.
       build_file = str(Label("//dotnet:nunit.BUILD")),
   )
+
+  native.new_http_archive(
+      name = "nuget",
+      url = "https://github.com/mono/nuget-binary/archive/0811ba888a80aaff66a93a4c98567ce904ab2663.zip", # Sept 6, 2016
```
Comment (zaphar) [inline dotnet/csharp.bzl]: It would be awesome if we could use a bazel-mirror for this like we do for nunit.

File: dotnet/csharp.bzl [hunk 3]
```diff
@@ -431,48 +440,88 @@ def _nuget_package_impl(repository_ctx):
...
+    default=Label("@mono//bin:mono"),
+  ),
+  # Reference to the nuget.exe file
+  "nuget_exe":attr.label(
+    default=Label("@nuget//:nuget.exe"),
+  ),
+}
 
 # This rule is a repository rule and is only usable in WORKSPACE files.
 # due to some limitations of repository_rules it does require you to
```
Comment (zaphar) [inline dotnet/csharp.bzl]: We no longer require them to say where the nuget binary is now right? We should remove that from this comment.

Comment (pcj) [inline dotnet/csharp.bzl]: I'll clean up that comment, you are right.  I should also document the new rule as well.  Will fix that up.

**Comment by pcj** [comment]
Anything other requested changes on this PR?

**Pull request merged**